pallets/asset/src/identifier.rs

```diff
@@ -0,0 +1,176 @@
+use codec::{Decode, Encode};
+use core::convert::TryInto;
+use sp_std::vec::Vec;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum Identifier {
+    CUSIP([u8; 9]),
+    CINS([u8; 9]),
+    ISIN([u8; 12]),
+    LEI([u8; 20]),
+    EMPTY,
+}
+
+impl Default for Identifier {
+    fn default() -> Self {
+        Identifier::EMPTY
+    }
+}
+
+impl Identifier {
+    pub fn cusip(bytes: [u8; 9]) -> Option<Identifier> {
+        (cusip_checksum(&bytes[..8]) == bytes[8] - b'0').then_some(Identifier::CUSIP(bytes))
+    }
+
+    pub fn cins(bytes: [u8; 9]) -> Option<Identifier> {
```

Comment:
These validations are only run if someone uses these functions to create the ENUM object. Someone can just do `Identifier::CUSIP($something_not_valid)` to create the ENUM. In fact, the extrinsics directly create the enum objects so they bypass these validations entirely.

One way to fix this will be to implement a "validate" function that takes the enum as input and returns validation redult. We can then call it over all the identifiers in the asset module.

Reply:
Unfortunately, you can't make enum variants private so we are kind of out of luck. I wouldn't mind the "validate" function solution so let's do it.

Reply:
You cannot make enum variants private, but you can make them contain a type which the outside cannot construct or pattern match, e.g.,
```rust
#![allow(warnings)]

mod foo {
    pub enum Foo {
        Baz(Qux),
    }
    
    pub struct Qux { wibble: u8 }
}

use foo::*;

fn attempt(foo: Foo) {
    let Foo::Baz(Qux { wibble }) = foo;
}
```

=>

```
error[E0451]: field `wibble` of struct `foo::Qux` is private
  --> src/lib.rs:14:24
   |
14 |     let Foo::Baz(Qux { wibble }) = foo;
   |                        ^^^^^^ private field
```

Reply:
@Centril That might not work in our case since the object is technically created outside rust and passed to us after being serialized. We deserialize it in the extrinsic to get the object back. I am not sure if serde/parity-scale-codec will stop deserializing an object with a private field. We can do a test but I think it's just better to call the validate function. We somehow need the users to pass data to us anyway and we can't be sure that the correct data is being sent without actually checking it.

Reply:
Ostensibly, since `#[derive(De/Serialize, En/Decode)]` is done in the same module where the type is defined, they should have access to the fields both for pattern matching & construction. Do we want to prevent that from happening?

Reply:
Not really, we do need it to be de/serializeable to read/write it to storage. What we want to prevent is end-users submitting objects that do not conform to the validation checks we have added. I think the easiest solution is to just implement a validate function and call it every time we take input from the user. Open to ideas.

Reply:
I pushed a validate func let me know how it looks